test/integration/fdp-class.browser.spec.ts

```diff
@@ -11,105 +31,388 @@ describe('Fair Data Protocol class - in browser', () => {
 
-        return {
-          beeUrl: fdp.connection.bee.url,
-          beeDebugUrl: fdp.connection.beeDebug.url,
-        }
-      },
-      BEE_URL,
-      BEE_DEBUG_URL,
-    )
+      return {
+        beeUrl: fdp.connection.bee.url,
+        beeDebugUrl: fdp.connection.beeDebug.url,
+      }
+    })
 
-    expect(urls.beeUrl).toBe(BEE_URL)
-    expect(urls.beeDebugUrl).toBe(BEE_DEBUG_URL)
+    expect(urls.beeUrl).toBe('http://localhost:1633')
+    expect(urls.beeDebugUrl).toBe('http://localhost:1635')
   })
 
   describe('Registration', () => {
-    it('register required users', async () => {
-      const usersList = [generateUser(), generateUser()]
-      const createdUsers = await page.evaluate(
-        async (BEE_URL, BEE_DEBUG_URL, users) => {
-          users = JSON.parse(users)
-          const fdp = new window.fdp.FairDataProtocol(BEE_URL, BEE_DEBUG_URL)
-          const result = []
-          for (const user of users) {
-            const data = await fdp.account.register(user.username, user.password, user.mnemonic)
-            result.push(data)
-
-            await fdp.account.import(user.username, user.mnemonic)
-            await fdp.account.login(user.username, user.password)
-          }
+    it('should register users', async () => {
+      const fairos = createFairosJs()
 
-          return result
-        },
-        BEE_URL,
-        BEE_DEBUG_URL,
-        JSON.stringify(usersList),
-      )
+      const usersList = [generateUser(), generateUser()] as unknown as JSONArray
+      const createdUsers = await page.evaluate(async users => {
+        const fdp = eval(await window.initFdp()) as FairDataProtocol
 
-      for (const [index, createdUser] of createdUsers.entries()) {
-        const user = usersList[index]
+        const result = []
+        for (const user of users) {
+          const data = await fdp.account.register(user.username, user.password, user.mnemonic)
+          result.push(data)
+
+          await fdp.account.import(user.username, user.mnemonic)
+          await fdp.account.login(user.username, user.password)
+        }
+
+        return result
+      }, usersList)
+
+      let i = 0
+      for (const createdUser of createdUsers) {
+        const user = usersList[i] as unknown as TestUser
         expect(createdUser.mnemonic).toEqual(user.mnemonic)
         expect(createdUser.wallet.address).toEqual(user.address)
         expect(createdUser.encryptedMnemonic).toBeDefined()
         expect(createdUser.reference).toBeDefined()
+        await fairos.userImport(user.username, user.password, '', user.address)
+        await fairos.userLogin(user.username, user.password)
+        i++
       }
     })
 
-    it('register already registered user', async () => {
+    it('should throw when registering already registered user', async () => {
+      await page.evaluate(async (user: TestUser) => {
+        const fdp = eval(await window.initFdp()) as FairDataProtocol
+        eval(await window.shouldFailString())
```

Comment:
In the context of the test, a test user is generated and passed to the context of the code in the web page.
In the context of the web page, the FDP and the error checking method are initialized. The initialization of these two methods happens via eval and strings, since there is no way to pass a function from the test context to the web page context without duplicating code. Only serializable data can be passed between contexts.
Then the user is registered, removed from the mapping, and the error is checked.

Reply:
now I see where you initialize but thanks for the description : )